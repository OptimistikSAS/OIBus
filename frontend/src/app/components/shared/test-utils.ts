import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
