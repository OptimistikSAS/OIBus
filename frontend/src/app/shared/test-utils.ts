import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
