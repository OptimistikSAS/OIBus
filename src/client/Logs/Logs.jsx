import React from 'react'
import { Link } from 'react-router-dom'
import {
  FormGroup,
  FormText,
  Label,
  Button,
  Input,
  Table,
  Card,
  CardBody,
  Row,
  Col,
  Breadcrumb,
  BreadcrumbItem,
} from 'reactstrap'
import apis from '../services/apis'
import { AlertContext } from '../context/AlertContext.jsx'

const selectStyle = (level) => {
  switch (level) {
    case 'debug':
      return 'text-success'
    case 'info':
      return 'text-info'
    case 'warning':
      return 'text-warning'
    case 'error':
      return 'text-danger'
    default:
      return 'text-primary'
  }
}

const Log = () => {
  const verbosityOptions = ['debug', 'info', 'warning', 'error', 'silly']
  const defaultMaxLog = 300
  const [fromDate, setFromDate] = React.useState()
  const [toDate, setToDate] = React.useState()
  const [verbosity, setVerbosity] = React.useState(verbosityOptions)
  const [logs, setLogs] = React.useState()
  const [filterText, setFilterText] = React.useState('')
  const [maxLog, setMaxLog] = React.useState(defaultMaxLog)
  const { setAlert } = React.useContext(AlertContext)

  /**
   * Handles the form's submittion and set the logs if any response
   * @returns {void}
   */
  const handleSubmit = async () => {
    try {
      const logsResponse = await apis.getLogs(fromDate, toDate, verbosity.join(','))
      // sort logs based on timestamp
      logsResponse.sort((a, b) => {
        if (a.timestamp < b.timestamp) {
          return 1
        }
        if (b.timestamp < a.timestamp) {
          return -1
        }
        return 0
      })
      setLogs(logsResponse)
      setMaxLog(defaultMaxLog)
    } catch (error) {
      console.error(error)
      setAlert({ text: error.message, type: 'danger' })
    }
  }

  /**
   * Handles the form's submittion and set the logs if any response
   * @param {string} value checked/unchecled item
   * @returns {void}
   */
  const handleVerbosityChange = (value) => {
    if (verbosity.includes(value)) {
      setVerbosity(verbosity.filter((item) => item !== value))
    } else {
      setVerbosity([...verbosity, value])
    }
  }

  /**
   * Render the filter options for logs
   * @returns {void}
   */
  const renderFilter = () => (
    <FormGroup row>
      <Label sm={1} for="filterText">
        Filter
      </Label>
      <Col sm={11}>
        <Input
          className="oi-form-input"
          type="text"
          id="filterText"
          placeholder="message contains..."
          value={filterText}
          onEnded={() => null}
          onChange={(event) => {
            setMaxLog(defaultMaxLog)
            setFilterText(event.target.value.toLowerCase())
          }}
        />
      </Col>
    </FormGroup>
  )

  /**
   * Render the logs
   * @returns {void}
   */
  const renderLogs = () => {
    const filteredLogs = logs.filter((item) => item.message.toLowerCase().includes(filterText))
    return (
      <Col className="log-right-panel">
        <Card>
          <Label className="label-card-title">Logs</Label>
          <CardBody className="card-body">
            {renderFilter()}
            <Table size="small" bordered hover responsive>
              <tbody>
                {filteredLogs
                  .filter((_, index) => index < maxLog)
                  .map((item) => {
                    const { id, source, level, message, timestamp } = item
                    const date = new Date(timestamp)
                    return (
                      <tr key={id} className={`oi-log ${selectStyle(level)}`}>
                        <td style={{ width: 120 }}>{`${date.toLocaleDateString()}-${date.toLocaleTimeString()}`}</td>
                        <td>{source}</td>
                        <td>{level}</td>
                        <td>{message}</td>
                      </tr>
                    )
                  })}
              </tbody>
            </Table>
            {filteredLogs.length > maxLog && (
              <Button color="primary" onClick={() => setMaxLog(maxLog + defaultMaxLog)}>
                Show more...
              </Button>
            )}
          </CardBody>
        </Card>
      </Col>
    )
  }

  const maxDateString = new Date().toISOString().substr(0, 16)
  return (
    <>
      <Breadcrumb tag="h5">
        <BreadcrumbItem tag={Link} to="/" className="oi-breadcrumb">
          Home
        </BreadcrumbItem>
        <BreadcrumbItem active tag="span">
          Logs
        </BreadcrumbItem>
      </Breadcrumb>
      <Row>
        <Col className="log-left-panel">
          <Card>
            <Label className="label-card-title">Filters</Label>
            <CardBody className="card-body">
              <FormGroup>
                <Label for="fromDate">From date</Label>
                <Input
                  className="oi-form-input"
                  type="datetime-local"
                  id="fromDatee"
                  max={maxDateString}
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required
                  onChange={(event) => {
                    const date = new Date(event.target.value).toISOString()
                    setFromDate(date)
                  }}
                />
                <FormText color="muted">default: Current datetime - 24 hours (now - 24 hours)</FormText>
              </FormGroup>

              <FormGroup>
                <Label for="toDate">To date</Label>
                <Input
                  className="oi-form-input"
                  type="datetime-local"
                  id="toDate"
                  max={maxDateString}
                  placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
                  required
                  onChange={(event) => {
                    const date = new Date(event.target.value).toISOString()
                    setToDate(date)
                  }}
                />
                <FormText color="muted">default: Current datetime (now)</FormText>
              </FormGroup>

              <FormGroup>
                <Label for="verbosity">Verbosity</Label>
                {verbosityOptions.map((item) => (
                  <FormGroup check className={`${selectStyle(item)}`} key={item}>
                    <Label check>
                      <Input
                        className="oi-form-input"
                        type="checkbox"
                        id="verbosity"
                        onChange={() => handleVerbosityChange(item)}
                        checked={verbosity.includes(item)}
                      />
                      {item}
                    </Label>
                  </FormGroup>
                ))}
              </FormGroup>
              <Button color="primary" onClick={handleSubmit}>
                Show log
              </Button>
            </CardBody>
          </Card>
        </Col>
        {logs && renderLogs()}
      </Row>
    </>
  )
}
export default Log
