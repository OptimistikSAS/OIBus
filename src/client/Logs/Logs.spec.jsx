import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'
import { BrowserRouter } from 'react-router-dom'

import Logs from './Logs.jsx'

// fixing date to match snapshot
const RealDate = Date
const realToLocaleString = global.Date.prototype.toLocaleString
const realToLocaleTimeString = global.Date.prototype.toLocaleTimeString
const realToLocaleDateString = global.Date.prototype.toLocaleDateString
const constantDate = new Date(Date.UTC(2020, 1, 1, 0, 0, 0))
// Ensure test output is consistent across machine locale and time zone config.
const mockToLocaleString = () => constantDate.toUTCString()

global.Date.prototype.toLocaleString = mockToLocaleString
global.Date.prototype.toLocaleTimeString = mockToLocaleString
global.Date.prototype.toLocaleDateString = mockToLocaleString

// Prevent random and time elements from failing repeated tests.
global.Date = class {
  static now() {
    return new RealDate(constantDate)
  }

  constructor() {
    return new RealDate(constantDate)
  }
}

// sample test logs
const testLogs = [
  {
    id: 1,
    timestamp: '2020-01-01T00:00:00.000Z',
    level: 'debug',
    source: 'Source',
    message: 'Testing debug logs',
  },
  {
    id: 2,
    timestamp: '2020-01-01T00:00:00.000Z',
    level: 'info',
    source: 'Source',
    message: 'Testing info logs',
  },
  {
    id: 3,
    timestamp: '2020-01-01T00:00:00.000Z',
    level: 'warning',
    source: 'Source',
    message: 'Testing warning logs',
  },
  {
    id: 4,
    timestamp: '2020-01-01T00:00:00.000Z',
    level: 'error',
    source: 'Source',
    message: 'Testing error logs',
  },
  {
    id: 5,
    timestamp: '2020-01-01T00:00:00.000Z',
    level: 'silly',
    source: 'Source',
    message: 'Testing silly logs',
  },
]

// setting initial test logs to be displayed
const setState = jest.fn()
React.useState = jest.fn().mockImplementation((init) => [init === undefined ? testLogs : init, setState])

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('Logs', () => {
  test('check Logs', () => {
    act(() => {
      ReactDOM.render(
        <BrowserRouter>
          <Logs />
        </BrowserRouter>,
        container,
      )
    })
    expect(container).toMatchSnapshot()
    global.Date = RealDate
    global.Date.prototype.toLocaleString = realToLocaleString
    global.Date.prototype.toLocaleTimeString = realToLocaleTimeString
    global.Date.prototype.toLocaleDateString = realToLocaleDateString
  })
})
