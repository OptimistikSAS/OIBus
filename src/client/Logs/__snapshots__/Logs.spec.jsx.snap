// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs check Logs 1`] = `
<div>
  <h5
    aria-label="breadcrumb"
    class=""
  >
    <ol
      class="breadcrumb"
    >
      <a
        class="oi-breadcrumb breadcrumb-item"
        href="/"
      >
        Home
      </a>
      <span
        aria-current="page"
        class="active breadcrumb-item"
      >
        Logs
      </span>
    </ol>
  </h5>
  <div
    class="row"
  >
    <div
      class="log-left-panel col"
    >
      <div
        class="card"
      >
        <label
          class="label-card-title"
        >
          Filters
        </label>
        <div
          class="card-body card-body"
        >
          <div
            class="form-group"
          >
            <label
              class=""
              for="fromDate"
            >
              From date
            </label>
            <input
              class="oi-form-input form-control"
              id="fromDatee"
              max="2020-02-01T00:00"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
            />
            <small
              class="form-text text-muted"
            >
              default: Current datetime - 24 hours (now - 24 hours)
            </small>
          </div>
          <div
            class="form-group"
          >
            <label
              class=""
              for="toDate"
            >
              To date
            </label>
            <input
              class="oi-form-input form-control"
              id="toDate"
              max="2020-02-01T00:00"
              placeholder="yyyy-mm-ddThh:mm:ss+hh:mm"
              required=""
              type="datetime-local"
            />
            <small
              class="form-text text-muted"
            >
              default: Current datetime (now)
            </small>
          </div>
          <div
            class="form-group"
          >
            <label
              class=""
              for="verbosity"
            >
              Verbosity
            </label>
            <div
              class="text-success form-check"
            >
              <label
                class="form-check-label"
              >
                <input
                  checked=""
                  class="oi-form-input form-check-input"
                  id="verbosity"
                  type="checkbox"
                />
                debug
              </label>
            </div>
            <div
              class="text-info form-check"
            >
              <label
                class="form-check-label"
              >
                <input
                  checked=""
                  class="oi-form-input form-check-input"
                  id="verbosity"
                  type="checkbox"
                />
                info
              </label>
            </div>
            <div
              class="text-warning form-check"
            >
              <label
                class="form-check-label"
              >
                <input
                  checked=""
                  class="oi-form-input form-check-input"
                  id="verbosity"
                  type="checkbox"
                />
                warning
              </label>
            </div>
            <div
              class="text-danger form-check"
            >
              <label
                class="form-check-label"
              >
                <input
                  checked=""
                  class="oi-form-input form-check-input"
                  id="verbosity"
                  type="checkbox"
                />
                error
              </label>
            </div>
            <div
              class="text-primary form-check"
            >
              <label
                class="form-check-label"
              >
                <input
                  checked=""
                  class="oi-form-input form-check-input"
                  id="verbosity"
                  type="checkbox"
                />
                silly
              </label>
            </div>
          </div>
          <button
            class="btn btn-primary"
            type="button"
          >
            Show log
          </button>
        </div>
      </div>
    </div>
    <div
      class="log-right-panel col"
    >
      <div
        class="card"
      >
        <label
          class="label-card-title"
        >
          Logs
        </label>
        <div
          class="card-body card-body"
        >
          <div
            class="row form-group"
          >
            <label
              class="col-sm-1 col-form-label"
              for="filterText"
            >
              Filter
            </label>
            <div
              class="col-sm-11"
            >
              <input
                class="oi-form-input form-control"
                id="filterText"
                placeholder="message contains..."
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="table-responsive"
          >
            <table
              class="table table-small table-bordered table-hover"
            >
              <tbody>
                <tr
                  class="oi-log text-success"
                >
                  <td
                    style="width: 120px;"
                  >
                    Sat, 01 Feb 2020 00:00:00 GMT-Sat, 01 Feb 2020 00:00:00 GMT
                  </td>
                  <td>
                    Source
                  </td>
                  <td>
                    debug
                  </td>
                  <td>
                    Testing debug logs
                  </td>
                </tr>
                <tr
                  class="oi-log text-info"
                >
                  <td
                    style="width: 120px;"
                  >
                    Sat, 01 Feb 2020 00:00:00 GMT-Sat, 01 Feb 2020 00:00:00 GMT
                  </td>
                  <td>
                    Source
                  </td>
                  <td>
                    info
                  </td>
                  <td>
                    Testing info logs
                  </td>
                </tr>
                <tr
                  class="oi-log text-warning"
                >
                  <td
                    style="width: 120px;"
                  >
                    Sat, 01 Feb 2020 00:00:00 GMT-Sat, 01 Feb 2020 00:00:00 GMT
                  </td>
                  <td>
                    Source
                  </td>
                  <td>
                    warning
                  </td>
                  <td>
                    Testing warning logs
                  </td>
                </tr>
                <tr
                  class="oi-log text-danger"
                >
                  <td
                    style="width: 120px;"
                  >
                    Sat, 01 Feb 2020 00:00:00 GMT-Sat, 01 Feb 2020 00:00:00 GMT
                  </td>
                  <td>
                    Source
                  </td>
                  <td>
                    error
                  </td>
                  <td>
                    Testing error logs
                  </td>
                </tr>
                <tr
                  class="oi-log text-primary"
                >
                  <td
                    style="width: 120px;"
                  >
                    Sat, 01 Feb 2020 00:00:00 GMT-Sat, 01 Feb 2020 00:00:00 GMT
                  </td>
                  <td>
                    Source
                  </td>
                  <td>
                    silly
                  </td>
                  <td>
                    Testing silly logs
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
